Source/DafnyCore/Resolver.cs

```diff
@@ -9944,43 +9944,109 @@ public CheckDividedConstructorInit_Visitor(Resolver resolver)
       return true;
     }
 
-    bool CheckIfEqualityIsDefinitelyNotSupported(IndDatatypeDecl dt, Graph<IndDatatypeDecl/*!*/>/*!*/ dependencies) {
+    /// <summary>
+    /// The following two functions `BasicCheckIfEqualityIsDefinitelyNotSupported` and `CheckIfEqualityIsDefinitelyNotSupported`
+    /// are mutually recursive and their purpose is to determine if a datatype cannot possibly support equality.
+    /// The entry point is always functoin `CheckIfEqualityIsDefinitelyNotSupported`
+    /// Function `CheckIfEqualityIsDefinitelyNotSupported` checks datatypes specifically
+    /// Function `BasicCheckIfEqualityIsDefinitelyNotSupported` cheks all other types
+    /// 
+    /// As a first approximation, a datatype definely cannot support equality if any of the following holds:
+    /// Any of its constructors is ghost
+    /// Any of the formal parameters of Any of its constructors is ghost
+    /// It is defined in terms of another type that definitely does not support equality, that is:
+    /// A datatype that does not support equality
+    /// An opaque type that has not been declared as supporting equality
+    /// A codatatype
+    /// An arrow type
```

Comment:
Given this list, shouldn't definitely supporting equality or not be computed by a member function defined in the `Type` interface, rather than a recursive function with type cases?

Reply:
I don't know. I like the way it is written now because the code almost reads as a spec, but I assume it is not idiomatic. I will leave it to you to decide what style you prefer.